### ISSUE TITLE
Replying to a mailing list with Umlaut in name fails

### ISSUE THREAD

**Issue by varac** [pazz/alot#1235]
Software Versions
-----------------

Python version: 2.7.14

Notmuch version: 0.26

Alot version (either a number or sha for master): 0.7

Observed behavior
-----------------

I want to reply to a mail that was send to a mailing list, where the name of the list (not the email adr) contains a german Umlaut:

```
From    ...
To      Möwe <moewe@gmail.com>
```

Alot refuses to do so, failing with this traceback:

```
DEBUG:ui:Got key (['r'], [114])
DEBUG:ui:cmdline: 'reply --all'
DEBUG:ui:thread command string: "reply --all"
DEBUG:__init__:mode:thread got commandline "reply --all"
DEBUG:__init__:ARGS: [u'reply', u'--all']
DEBUG:__init__:cmd parms {'spawn': None, 'all': True, 'listreply': None}
DEBUG:envelope:TEMPLATE: None
DEBUG:utils:unquoted header: |[MA] heute|
DEBUG:thread:candidate addresses: [('=?ISO-8859-1?Q?M=F6we?=', 'moewe@gmail.com')]
…
DEBUG:thread:using realname: "Varac"
DEBUG:thread:using address: varac@…
DEBUG:thread:reply to: =?ISO-8859-1?Q?M=F6we?= <moewe@gmail.com>
DEBUG:utils:unquoted header: |=?ISO-8859-1?Q?M=F6we?= <moewe@gmail.com>
DEBUG:globals:Trying to encrypt message because encrypt=False and encrypt_by_default=all
DEBUG:utils:encryption keys: [u'M\xf6we Altonah', u'moewe@gmail.com']
ERROR:ui:Traceback (most recent call last):
  File "/home/varac/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1532, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
  File "/home/varac/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/varac/.local/lib/python2.7/site-packages/alot-0.7-py2.7.egg/alot/commands/utils.py", line 47, in set_encrypt
    signed_only=signed_only)
  File "/home/varac/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1532, in unwindGenerator
    return _inlineCallbacks(None, gen, Deferred())
--- <exception caught here> ---
  File "/home/varac/.local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/varac/.local/lib/python2.7/site-packages/alot-0.7-py2.7.egg/alot/commands/utils.py", line 90, in _get_keys
    signed_only=signed_only)
  File "/home/varac/.local/lib/python2.7/site-packages/alot-0.7-py2.7.egg/alot/crypto.py", line 72, in get_key
    raise GPGProblem('Cannot find key for "{}".'.format(keyid),
exceptions.UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 1: ordinal not in range(128)
```

Expected behavior
-----------------

I'd like to reply this mail without any issue.

**Comment by dkg** [comment]
On Sun 2018-04-15 08:41:19 +0000, Varac wrote:
> DEBUG:thread:candidate addresses: [('=?ISO-8859-1?Q?M=F6we?=', 'moewe@gmail.com')]

this is problematic -- why is it choosing ISO-8859-1 instead of UTF-8?

> DEBUG:utils:encryption keys: [u'M\xf6we Altonah', u'moewe@gmail.com']

and this is the consequence -- after inverting the RFC 2047 encoding, we
get an ISO-8859-1 bytestream back.  it's inappropriate to treat the
bytestream as UTF-8 directly, so the "u" prefix above doesn't make
sense.

> exceptions.UnicodeEncodeError: 'ascii' codec can't encode character u'\xf6' in position 1: ordinal not in range(128)

and this is the error message where things finally break.

So i think the right solution is several-fold:

 * if we're generating RFC 2047-encoded subject parts for non-ascii
   text, we should just use UTF-8, not any other character encoding.  it
   makes everyone's life easier.  it's 2018!

 * if we're just copy/pasting or replaying subject parts (or other
   headers) that have been RFC 2047-encoded using some non-UTF-8
   character encoding, then whenever we try to reuse them internally, we
   need to:

    a) strip the 2047-encoding to get a bytestream and a character
       encoding.

    b) use the character encoding to convert the bytestream to a unicode
       string.

**Comment by varac** [comment]
Btw, I get a similar error when trying to attach files with german Umlauts as well.

**Comment by pazz** [comment]
Can you confirm that this is still an issue on python3 (master) ?

**Comment by varac** [comment]
I gave latest master a try but hit #1246, besides other issues. I'd like to wait for a more stable release to test this.

**Comment by pazz** [comment]
Quoting Varac (2018-06-20 14:46:29)
> I gave latest master a try but hit #1246, besides other issues. I'd like to
> wait for a more stable release to test this.

If you're seeing the error message mention in the initial report of #1246:

```
...
from ..settings.const import settings
ImportError: cannot import name 'settings'
```

Then this is actually expected! This should never have worked in the first place and should be fixed by #1251.
Please report errors that you observe on the master branch unless they are already documented.

**Comment by varac** [comment]
I gave it another try (based on https://github.com/pazz/alot/pull/1251) and now I cannot reproduce this issue anymore. Nice!
I'm fine with closing this for now.
